src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,106 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_INDEX = 10;
+    private static final double GROW_INDEX = 1.5;
+    private int maxSize = DEFAULT_INDEX;
+    private int sizeOfArray;
+    private T[] array;
+
+    public ArrayList() {
+        array = (T[]) new Object[maxSize];
+    }
+
     @Override
     public void add(T value) {
-
+        resize();
+        array[sizeOfArray] = value;
+        sizeOfArray++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index > sizeOfArray || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("No such index exception: "
+                    + index + " size " + sizeOfArray);
+        }
+        resize();
+        if (index < sizeOfArray) {
+            System.arraycopy(array, index, array, index + 1, sizeOfArray - index);
+            array[index] = value;
+            sizeOfArray++;
+            return;
+        }
+        array[index] = value;
+        sizeOfArray++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            T value = list.get(i);
+            add(value);
+        }
     }
 
     @Override
-    public T get(int index) {
-        return null;
+    public T get(int index) throws ArrayListIndexOutOfBoundsException {
+        errorCheck(index);
+        return array[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        errorCheck(index);
+        array[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        errorCheck(index);
+        T result = array[index];
+        System.arraycopy(array, index + 1, array, index, sizeOfArray - index - 1);
+        sizeOfArray--;
+        return result;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        for (int i = 0; i < sizeOfArray; i++) {
+            if ((element != null && element.equals(array[i]))
+                    || element == array[i]) {
+                remove(i);
+                return element;
+            }
+        }
+        throw new NoSuchElementException("No such index exception: " + element + " is absent");
     }
 
     @Override
     public int size() {
-        return 0;
+        return sizeOfArray;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return sizeOfArray == 0;
+    }
+
+    private void resize() {
+        if (sizeOfArray == maxSize) {
+            maxSize = (int) (maxSize * GROW_INDEX);
+            T[] arrayTemporary = (T[]) new Object[maxSize];
+            System.arraycopy(array, 0, arrayTemporary, 0, sizeOfArray);
+            array = arrayTemporary;
+        }
+    }
+
+    private void errorCheck(int index) {
```

Comment:
method name doesn't make it clear what it checks, think of a better one, previous reviewer proposed `rangeCheck`  for example